core/src/main/java/com/linecorp/armeria/internal/server/CorsHeaderUtil.java

```diff
@@ -0,0 +1,209 @@
+    }
+
+    public static void setCorsAllowHeaders(RequestHeaders requestHeaders, ResponseHeadersBuilder headers,
+                                           CorsPolicy corsPolicy) {
+        if (corsPolicy.shouldAllowAllRequestHeaders()) {
+            final String header = requestHeaders.get(HttpHeaderNames.ACCESS_CONTROL_REQUEST_HEADERS);
+            if (!Strings.isNullOrEmpty(header)) {
+                headers.set(HttpHeaderNames.ACCESS_CONTROL_ALLOW_HEADERS, header);
+            }
+
+            return;
+        }
+
+        if (corsPolicy.allowedRequestHeaders().isEmpty()) {
+            return;
+        }
+
+        headers.set(HttpHeaderNames.ACCESS_CONTROL_ALLOW_HEADERS, joinAllowedRequestHeaders(corsPolicy));
+    }
+
+    /**
+     * Sets origin header according to the given CORS configuration and HTTP request.
+     *
+     * @param request the HTTP request
+     * @param headers the HTTP headers to modify
+     *
+     * @return {@code policy} if CORS configuration matches, otherwise {@code null}
+     */
+    @Nullable
+    public static CorsPolicy setCorsOrigin(ServiceRequestContext ctx, HttpRequest request,
+                                           ResponseHeadersBuilder headers, CorsConfig config) {
+
+        final String origin = request.headers().get(HttpHeaderNames.ORIGIN);
+        if (origin != null) {
+            final CorsPolicy policy = config.getPolicy(origin, ctx.routingContext());
+            if (policy == null) {
+                logger.debug(
+                        "{} There is no CORS policy configured for the request origin '{}' and the path '{}'.",
+                        ctx, origin, ctx.path());
+                return null;
+            }
+            if (NULL_ORIGIN.equals(origin)) {
+                setCorsNullOrigin(headers);
+                return policy;
+            }
+            if (config.isAnyOriginSupported()) {
+                if (policy.isCredentialsAllowed()) {
+                    echoCorsRequestOrigin(request, headers);
+                    setCorsVaryHeader(headers);
+                } else {
+                    setCorsAnyOrigin(headers);
+                }
+                return policy;
+            }
+            setCorsOrigin(headers, origin);
+            setCorsVaryHeader(headers);
+            return policy;
+        }
+        return null;
+    }
+
+    private static void setCorsOrigin(ResponseHeadersBuilder headers, String origin) {
+        headers.set(HttpHeaderNames.ACCESS_CONTROL_ALLOW_ORIGIN, origin);
+    }
+
+    private static void echoCorsRequestOrigin(HttpRequest request, ResponseHeadersBuilder headers) {
+        final String origin = request.headers().get(HttpHeaderNames.ORIGIN);
+        if (origin != null) {
+            setCorsOrigin(headers, origin);
+        }
+    }
+
+    private static void setCorsVaryHeader(ResponseHeadersBuilder headers) {
+        headers.set(HttpHeaderNames.VARY, HttpHeaderNames.ORIGIN.toString());
+    }
```

Comment:
What do you think of adding instead of setting because there might be other vary headers set already?
```suggestion
    private static void addCorsVaryHeader(ResponseHeadersBuilder headers) {
        headers.add(HttpHeaderNames.VARY, HttpHeaderNames.ORIGIN.toString());
    }
```